docker/go-control-plane/fallback-control-plane.go

```diff
@@ -0,0 +1,207 @@
+// the resource being updated.
+func (srv *controlService) UpsertResources(_ context.Context, req *cs.UpsertResourcesRequest) (*cs.UpsertResourcesResponse, error) {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	srv.version++
+	listener := controlplane.ListenerName
+	if req.Listener != nil {
+		listener = *req.Listener
+	}
+	srv.clusters[req.Cluster] = controlplane.MakeCluster(req.Cluster, req.UpstreamHost, req.UpstreamPort)
+	srv.listeners[listener] = controlplane.MakeHTTPListener(listener, req.Cluster)
+	snapshot, err := srv.MakeSnapshot()
+	if err != nil {
+		l.Errorf("snapshot inconsistency: %+v\n", err)
+		return nil, err
+	}
+	srv.cache.SetSnapshot(context.Background(), *nodeID, snapshot)
+	res := &cs.UpsertResourcesResponse{}
+	for _, l := range srv.listeners {
+		a, err := anypb.New(l)
+		if err != nil {
+			panic(err)
+		}
+		res.Resource = append(res.Resource, a)
+	}
+	for _, c := range srv.clusters {
+		a, err := anypb.New(c)
+		if err != nil {
+			panic(err)
+		}
+		res.Resource = append(res.Resource, a)
+	}
+	return res, nil
+}
+
+// MakeSnapshot builds xDS configuration snapshot that will be served
+// to clients.
+func (srv *controlService) MakeSnapshot() (*cache.Snapshot, error) {
+	listeners := make([]types.Resource, len(srv.listeners))
+	i := 0
+	for _, l := range srv.listeners {
+		listeners[i] = l
+		i++
+	}
+	clusters := make([]types.Resource, len(srv.clusters))
+	i = 0
+	for _, c := range srv.clusters {
+		clusters[i] = c
+		i++
+	}
+	resources := map[resource.Type][]types.Resource{resource.ListenerType: listeners, resource.ClusterType: clusters}
+	// Create the snapshot that we'll serve to Envoy
+	snapshot, error := cache.NewSnapshot(fmt.Sprint(srv.version), resources)
+	if error != nil {
+		return nil, error
+	}
+	if err := snapshot.Consistent(); err != nil {
+		l.Errorf("snapshot inconsistency: %+v\n", err)
+		for _, r := range snapshot.Resources {
+			for name, resource := range r.Items {
+				bytes, err := prototext.MarshalOptions{Multiline: true}.Marshal(resource.Resource)
+				if err != nil {
+					l.Errorf("Can't marshal %s\n", name)
+				} else {
+					l.Errorf("Resource: %s\n%s\n",
+						resource.Resource,
+						string(bytes))
+				}
+			}
+		}
+		return nil, err
+	}
+	l.Debugf("will serve snapshot:\n")
+	for _, values := range snapshot.Resources {
+		for name, item := range values.Items {
+			text, err := prototext.MarshalOptions{Multiline: true}.Marshal(item.Resource)
+			if err != nil {
+				l.Errorf("Resource %+v, error: %+v\n", name, err)
+			} else {
+				l.Debugf("%+v => %+v\n", name, string(text))
+			}
+		}
+	}
+	return snapshot, nil
+}
+
+// Main entry point. Configures and starts a gRPC server that serves xDS traffic
+// and provides an interface for tests to manage control plane behavior.
+func main() {
+	flag.Parse()
+	sep := strings.LastIndex(*upstream, ":")
+	if sep < 0 {
+		l.Errorf("Incorrect upstream host name: %+v\n", upstream)
+		os.Exit(1)
```

Comment:
Why do these log messages need to use the other logger? Why can't these simply use the `log` package? And if we used the `log` package, these two lines can be replaced with a call to `log.Exitf`. Please see: go/go-style/decisions#dont-panic